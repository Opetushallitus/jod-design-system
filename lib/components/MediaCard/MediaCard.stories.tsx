import type { Meta, StoryObj } from '@storybook/react';

import { useState } from '@storybook/preview-api';
import { fn } from '@storybook/test';
import { MediaCard } from './MediaCard';

const meta = {
  title: 'Cards/MediaCard',
  component: MediaCard,
  tags: ['autodocs'],
} satisfies Meta<typeof MediaCard>;

export default meta;

type Story = StoryObj<typeof meta>;
const imageSrc = 'https://images.unsplash.com/photo-1523464862212-d6631d073194?q=80&w=260';

export const Default: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8529',
    },
    docs: {
      description: {
        story: 'Vertical MediaCard where there is no favorite-button available',
      },
    },
  },
  args: {
    imageSrc,
    imageAlt: 'Woman standing in front of a colourful wall',
    label: 'Tulevaisuusmatka',
    description: 'Tulevaisuusmatka on koulutus, joka auttaa sinua löytämään oman polkusi ja tavoitteesi.',
    tags: ['Taglorem', 'Loremtag', 'Nonutag'],
  },
};

export const Horizontal: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8529',
    },
    docs: {
      description: {
        story: 'Horizontal MediaCard where there is no favorite-button available',
      },
    },
  },
  args: {
    variant: 'horizontal',
    imageSrc,
    imageAlt: 'Woman standing in front of a colourful wall',
    label: 'Tulevaisuusmatka',
    description: 'Tulevaisuusmatka on koulutus, joka auttaa sinua löytämään oman polkusi ja tavoitteesi.',
    tags: ['Taglorem', 'Loremtag', 'Nonutag'],
  },
};

export const Multiple: Story = {
  render: (args) => {
    const longTitle = 'Otsikko joka aina vain jatkuu ja jatkuu ja rivittyy';
    const shortTitle = 'Lyhyt otsikko';
    const longDescription = 'Teksti joka on tosi pitkä, eikä se mahdu millään yhdelle tai edes kahdelle riville.';
    const shortDescription = 'Tosi vähän tekstiä';
    return (
      <div className="ds:flex ds:flex-row ds:gap-4">
        <MediaCard {...args} label={longTitle} description={longDescription} />
        <MediaCard {...args} label={longTitle} description={shortDescription} />
        <MediaCard {...args} label={shortTitle} description={longDescription} tags={args.tags.slice(0, 3)} />
        <MediaCard {...args} label={shortTitle} description={shortDescription} tags={[]} />
      </div>
    );
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8529',
    },
    docs: {
      description: {
        story: `Group of Vertical MediaCards with different lengths of title, description and tags.`,
      },
    },
  },
  args: {
    imageSrc,
    imageAlt: 'Woman standing in front of a colourful wall',
    label: 'Otsikko joka aina vain jatkuu ja jatkuu ja rivittyy',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam adipiscin amet consectetur.',
    tags: ['Taglorem', 'Loremtag', 'Nonutag', 'Dolortag', 'Tagamet', 'Nullatag'],
  },
};

const getAbleToBeFavorite = (description: string, variant: 'horizontal' | 'vertical'): Story => {
  return {
    parameters: {
      docs: {
        description: {
          story: description,
        },
      },
      design: {
        type: 'figma',
        url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-8529',
      },
    },
    render: (args) => {
      const [isFavorite, setFavorite] = useState(true);
      return (
        <MediaCard
          {...args}
          favoriteLabel={isFavorite ? 'Poista suosikeista' : 'Lisää suosikkeihin'}
          isFavorite={isFavorite}
          onFavoriteClick={() => setFavorite(!isFavorite)}
        />
      );
    },
    args: {
      variant,
      imageSrc,
      imageAlt: 'Woman standing in front of a colourful wall',
      label: 'Tulevaisuusmatka',
      description: 'Tulevaisuusmatka on koulutus, joka auttaa sinua löytämään oman polkusi ja tavoitteesi.',
      tags: ['Taglorem', 'Loremtag', 'Nonutag'],
      isFavorite: true,
      onFavoriteClick: fn(),
      favoriteLabel: 'Poista suosikeista',
    },
  };
};

export const AbleToBeFavoritedVertical: Story = getAbleToBeFavorite(
  'Vertical MediaCard where the favorite status can be toggled.',
  'vertical',
);

export const AbleToBeFavoritedHorizontal: Story = getAbleToBeFavorite(
  'Horizontal MediaCard where the favorite status can be toggled.',
  'horizontal',
);
