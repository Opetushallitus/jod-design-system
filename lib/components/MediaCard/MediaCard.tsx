import React from 'react';
import { MdFavorite, MdFavoriteBorder } from 'react-icons/md';
import { useMediaQueries } from '../../hooks/useMediaQueries';
import { LazyImage } from '../LazyImage/LazyImage';

type LinkComponent =
  | {
      to: string;
      linkComponent: React.ComponentType<{
        to: string;
        className?: string;
        children: React.ReactNode;
      }>;
    }
  | {
      to?: never;
      linkComponent?: never;
    };

type FavoriteButtonProps =
  | {
      isFavorite?: never;
      onFavoriteClick?: never;
      favoriteLabel?: never;
    }
  | {
      isFavorite: boolean;
      onFavoriteClick: () => void;
      favoriteLabel: string;
    };

export type MediaCardProps = {
  variant?: 'vertical' | 'horizontal';
  imageSrc: string;
  imageAlt: string;
  label: string;
  description: string;
  tags: {
    label: string;
    to: string;
  }[];
} & LinkComponent &
  FavoriteButtonProps;

type MediaCardImplProps = {
  imageSrc: string;
  imageAlt: string;
  label: string;
  description: string;
  children: React.ReactNode;
} & LinkComponent;

const Tag = ({
  label,
  to,
  linkComponent: Link,
}: {
  label: string;
  to: string;
  linkComponent?: React.ComponentType<{
    to: string;
    className?: string;
    children: React.ReactNode;
  }>;
}) => {
  return (
    <li className="ds:px-2 ds:pb-2 ds:last:pr-0 ds:first:pl-0">
      {Link ? (
        <Link to={to} className="ds:relative ds:z-1">
          {label}
        </Link>
      ) : (
        <>{label}</>
      )}
    </li>
  );
};

const LinkOrDiv = ({
  to,
  linkComponent: Link,
  className,
  children,
}: {
  to?: string;
  linkComponent?: React.ComponentType<{
    to: string;
    className?: string;
    children: React.ReactNode;
  }>;
  className?: string;
  children: React.ReactNode;
}) => {
  return Link && to ? (
    <Link
      to={to}
      className={`ds:z-1 ds:before:content-[''] ds:before:absolute ds:before:w-full ds:before:h-full ${className}`}
    >
      {children}
    </Link>
  ) : (
    <div className={`${className}`}>{children}</div>
  );
};

const FavoriteButton = ({ isFavorite, favoriteLabel, onFavoriteClick }: FavoriteButtonProps) => {
  return (
    <button
      className="ds:cursor-pointer ds:absolute ds:top-0 ds:right-0 ds:p-[12px] ds:bg-white ds:rounded-bl ds:z-1"
      aria-label={favoriteLabel}
      onClick={onFavoriteClick}
    >
      {isFavorite ? (
        <MdFavorite size={24} aria-hidden className="ds:text-accent" />
      ) : (
        <MdFavoriteBorder aria-hidden size={24} />
      )}
    </button>
  );
};

export const MediaCard = ({
  variant = 'vertical',
  imageSrc,
  imageAlt,
  label,
  description,
  tags,
  linkComponent: Link,
  to,
  isFavorite,
  onFavoriteClick,
  favoriteLabel,
}: MediaCardProps) => {
  const favoriteButtonAndTags = (
    <>
      {isFavorite !== undefined && (
        <FavoriteButton isFavorite={isFavorite} favoriteLabel={favoriteLabel} onFavoriteClick={onFavoriteClick} />
      )}
      <ul className="ds:text-attrib-value ds:flex ds:flex-row ds:divide-x ds:divide-secondary-5 ds:flex-wrap ds:text-accent ds:pt-3 ds:px-5">
        {tags.filter(Boolean).map((tag) => (
          <Tag key={tag.label} linkComponent={Link} {...tag} />
        ))}
      </ul>
    </>
  );
  const linkProps = to && Link ? { to, linkComponent: Link } : {};
  const MediaCardComponent = variant === 'vertical' ? MediaCardVertical : MediaCardHorizontal;

  return (
    <MediaCardComponent imageSrc={imageSrc} imageAlt={imageAlt} label={label} description={description} {...linkProps}>
      {favoriteButtonAndTags}
    </MediaCardComponent>
  );
};

const MediaCardVertical = ({
  imageSrc,
  imageAlt,
  label,
  description,
  linkComponent: Link,
  to,
  children,
}: MediaCardImplProps) => {
  const variantImageClassNames = 'ds:object-cover ds:h-[147px] ds:min-h-[147px]';
  const labelRef = React.useRef<HTMLDivElement>(null);
  const SINGLE_LINE_LABEL_HEIGHT = 27;
  const [lineClampClassNames, setLineClampClassNames] = React.useState('ds:line-clamp-3');

  React.useEffect(() => {
    const labelHeight = labelRef.current?.clientHeight ?? 0;
    if (labelHeight > SINGLE_LINE_LABEL_HEIGHT) {
      setLineClampClassNames('ds:line-clamp-2');
    } else {
      setLineClampClassNames('ds:line-clamp-3');
    }
  }, [label]);

  return (
    <div className="ds:relative ds:flex ds:flex-col ds:w-[261px] ds:min-h-[299px] ds:overflow-clip ds:rounded ds:shadow-border ds:bg-white ds:pb-5">
      <LinkOrDiv to={to} linkComponent={Link} className="ds:grow ">
        {imageSrc ? (
          <LazyImage className={`${variantImageClassNames}`} src={imageSrc} alt={imageAlt} />
        ) : (
          <span className={`ds:w-full ds:h-full ds:bg-secondary-5 ds:max-w-[265px] ${variantImageClassNames}`}></span>
        )}
        <div className="ds:px-5 ds:pt-4 ds:text-black ds:flex ds:flex-col ds:justify-between ds:h-full ds:flex-nowrap">
          <div className="ds:gap-3 ds:flex ds:flex-col">
            <div className="ds:text-heading-3-mobile ds:sm:text-heading-3 ds:line-clamp-2" ref={labelRef}>
              {label}
            </div>
            <div className={`ds:text-body-sm-mobile ds:sm:text-body-sm ${lineClampClassNames}`}>{description}</div>
          </div>
        </div>
      </LinkOrDiv>
      {children}
    </div>
  );
};

const MediaCardHorizontal = ({
  imageSrc,
  imageAlt,
  label,
  description,
  linkComponent: Link,
  to,
  children,
}: MediaCardImplProps) => {
  const { sm } = useMediaQueries();

  return (
    <div className="ds:relative ds:min-h-[137px] ds:w-full ds:overflow-clip ds:rounded ds:shadow-border ds:bg-white ds:grid ds:grid-cols-1 ds:sm:grid-cols-[193px_1fr] ds:lg:grid-cols-[255px_1fr] ds:grid-rows-[1fr_35px]">
      <LinkOrDiv
        to={to}
        linkComponent={Link}
        className="ds:flex ds:flex-row ds:col-start-1 ds:col-end-2 ds:sm:col-end-3 ds:row-start-1 ds:row-end-3"
      >
        <div className="ds:shrink-0">
          {imageSrc ? (
            sm && (
              <LazyImage
                className="ds:sm:w-[193px] ds:lg:w-[255px] ds:sm:min-w-full ds:sm:min-h-full ds:sm:h-0 ds:object-cover"
                src={imageSrc}
                alt={imageAlt}
              />
            )
          ) : (
            <div
              className={`ds:sm:w-[193px] ds:lg:w-[255px] ds:sm:min-w-full ds:sm:min-h-full ds:bg-secondary-5`}
            ></div>
          )}
        </div>
        <div className="ds:px-5 ds:pt-4 ds:pb-5 ds:text-black ds:flex ds:flex-col ds:justify-between ds:flex-nowrap">
          <div className="ds:gap-3 ds:flex ds:flex-col">
            <div className="ds:text-heading-3-mobile ds:sm:text-heading-3 ds:line-clamp-3 ds:sm:line-clamp-2">
              {label}
            </div>
            <div className="ds:text-body-sm-mobile ds:sm:text-body-sm ds:line-clamp-3 ds:sm:line-clamp-2">
              {description}
            </div>
          </div>
        </div>
      </LinkOrDiv>
      <div className="ds:col-start-1 ds:sm:col-start-2 ds:row-start-2">{children}</div>
    </div>
  );
};
