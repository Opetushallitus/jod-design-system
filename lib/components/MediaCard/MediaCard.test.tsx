import { render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { MediaCard } from './MediaCard';

const mockTags = [
  { label: 'tag1', to: '#' },
  { label: 'tag2', to: '#' },
];
describe('MediaCard', () => {
  it('renders the MediaCard component with vertical variant', () => {
    render(
      <MediaCard
        variant="vertical"
        imageSrc="vertical.jpg"
        imageAlt="Image for the vertical"
        label="Vertical"
        description="Vertical description"
        tags={mockTags}
      />,
    );

    expect(screen.getByAltText('Image for the vertical')).toBeInTheDocument();
    expect(screen.getByText('Vertical')).toBeInTheDocument();
    expect(screen.getByText('Vertical description')).toBeInTheDocument();
    expect(screen.getByText('tag1')).toBeInTheDocument();
    expect(screen.getByText('tag2')).toBeInTheDocument();
  });

  it('renders the MediaCard component with horizontal variant', () => {
    vi.mock('../../hooks/useMediaQueries', () => ({
      useMediaQueries: vi.fn().mockReturnValue({ sm: true, md: false, lg: false, xl: false }),
    }));
    render(
      <MediaCard
        variant="horizontal"
        imageSrc="horizontal.jpg"
        imageAlt="Image for the horizontal"
        label="Horizontal"
        description="Horizontal description"
        tags={mockTags}
      />,
    );

    expect(screen.getByAltText('Image for the horizontal')).toBeInTheDocument();
    expect(screen.getByText('Horizontal')).toBeInTheDocument();
    expect(screen.getByText('Horizontal description')).toBeInTheDocument();
    expect(screen.getByText('tag1')).toBeInTheDocument();
    expect(screen.getByText('tag2')).toBeInTheDocument();
  });

  it('renders the MediaCard component with default vertical variant', () => {
    const { container } = render(
      <MediaCard
        imageSrc="default.jpg"
        imageAlt="Image for default"
        label="Default"
        description="Default description"
        tags={mockTags}
      />,
    );

    expect(screen.getByAltText('Image for default')).toBeInTheDocument();
    expect(screen.getByText('Default')).toBeInTheDocument();
    expect(screen.getByText('Default description')).toBeInTheDocument();
    expect(screen.getByText('tag1')).toBeInTheDocument();
    expect(screen.getByText('tag2')).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });
});
