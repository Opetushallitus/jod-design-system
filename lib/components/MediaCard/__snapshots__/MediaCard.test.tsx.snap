// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MediaCard > renders the MediaCard component with default vertical variant 1`] = `
<div
  class="ds:overflow-clip ds:rounded ds:shadow-border ds:bg-white ds:relative ds:flex ds:flex-col ds:w-[261px] ds:min-h-[299px]"
>
  <img
    alt="Image for default"
    class="ds:object-cover ds:h-[147px]"
    src="default.jpg"
  />
  <div
    class="ds:px-5 ds:pt-4 ds:pb-5 ds:text-black ds:flex ds:flex-col ds:justify-between ds:h-full ds:flex-nowrap"
  >
    <div
      class="ds:gap-3 ds:flex ds:flex-col"
    >
      <div
        class="ds:text-heading-3-mobile ds:sm:text-heading-3 ds:line-clamp-2"
      >
        Default
      </div>
      <div
        class="ds:text-body-sm-mobile ds:sm:text-body-sm ds:line-clamp-3"
      >
        Default description
      </div>
    </div>
    <ul
      class="ds:text-attrib-value ds:flex ds:flex-row ds:divide-x ds:divide-secondary-5 ds:flex-wrap ds:text-accent ds:pt-3"
    >
      <li
        class="ds:px-2 ds:pb-2 ds:last:pr-0"
      >
        tag1
      </li>
      <li
        class="ds:px-2 ds:pb-2 ds:last:pr-0"
      >
        tag2
      </li>
    </ul>
  </div>
</div>
`;
